packages/block/src/header.ts

```diff
@@ -4,58 +4,156 @@ import {
-  public gasLimit!: Buffer
-  public gasUsed!: Buffer
-  public timestamp!: Buffer
-  public extraData!: Buffer
-  public mixHash!: Buffer
-  public nonce!: Buffer
-
-  readonly _common: Common
+export class Header {
+  public readonly _common: Common
+  public readonly parentHash: Buffer
+  public readonly uncleHash: Buffer
+  public readonly coinbase: Buffer
+  public readonly stateRoot: Buffer
+  public readonly transactionsTrie: Buffer
+  public readonly receiptTrie: Buffer
+  public readonly bloom: Buffer
+  public readonly difficulty: Buffer
+  public readonly number: Buffer
+  public readonly gasLimit: Buffer
+  public readonly gasUsed: Buffer
+  public readonly timestamp: Buffer
+  public readonly extraData: Buffer
+  public readonly mixHash: Buffer
+  public readonly nonce: Buffer
+
+  public static fromHeaderData(headerData: HeaderData, opts: BlockOptions = {}) {
+    const {
+      parentHash,
+      uncleHash,
+      coinbase,
+      stateRoot,
+      transactionsTrie,
+      receiptTrie,
+      bloom,
+      difficulty,
+      number,
+      gasLimit,
+      gasUsed,
+      timestamp,
+      extraData,
+      mixHash,
+      nonce,
+    } = headerData
+
+    return new Header(
+      parentHash ? checkBufferLength(toBuffer(parentHash), 32) : zeros(32),
+      uncleHash ? toBuffer(uncleHash) : KECCAK256_RLP_ARRAY,
+      coinbase ? checkBufferLength(toBuffer(coinbase), 20) : zeros(20),
+      stateRoot ? checkBufferLength(toBuffer(stateRoot), 32) : zeros(32),
+      transactionsTrie ? checkBufferLength(toBuffer(transactionsTrie), 32) : KECCAK256_RLP,
+      receiptTrie ? checkBufferLength(toBuffer(receiptTrie), 32) : KECCAK256_RLP,
+      bloom ? toBuffer(bloom) : zeros(256),
+      difficulty ? toBuffer(difficulty) : Buffer.from([]),
+      number ? toBuffer(number) : Buffer.from([]),
+      gasLimit ? toBuffer(gasLimit) : Buffer.from('ffffffffffffff', 'hex'),
+      gasUsed ? toBuffer(gasUsed) : Buffer.from([]),
+      timestamp ? toBuffer(timestamp) : Buffer.from([]),
+      extraData ? toBuffer(extraData) : Buffer.from([]),
+      mixHash ? toBuffer(mixHash) : zeros(32),
+      nonce ? toBuffer(nonce) : zeros(8),
+      opts,
+    )
+  }
+
+  public static fromRLPSerializedHeader(serialized: Buffer, opts: BlockOptions) {
+    const values = rlp.decode(serialized)
+
+    if (!Array.isArray(values)) {
+      throw new Error('Invalid serialized header input. Must be array')
+    }
+
+    return this.fromValuesArray(values, opts)
+  }
+
+  public static fromValuesArray(values: Buffer[], opts: BlockOptions) {
+    if (values.length > 15) {
+      throw new Error('invalid header. More values than expected were received')
+    }
+
+    const [
+      parentHash,
+      uncleHash,
+      coinbase,
+      stateRoot,
+      transactionsTrie,
+      receiptTrie,
+      bloom,
+      difficulty,
+      number,
+      gasLimit,
+      gasUsed,
+      timestamp,
+      extraData,
+      mixHash,
+      nonce,
+    ] = values
```

Comment:
Left this so expanded since I assume that these values need to be adopted (as in the tx PR) when moving over to domain specific value representations. So this structure will likely be needed anyhow.